### DISCUSSION TITLE
SIMD Best Practice for Vectorisation

### DISCUSSION THREAD

**Discussion by hwhitehead** [PrincetonUniversity/athena#538] [Q&A] [answered]
In attempt to accelerate my own custom sourcing, I have been referring to the use of SIMD vectorisation within Athena++

I have noted in multiple files the use of if statements within "#pragma omp simd" blocks, normally to include non barotropic effects .e.g with the NON_BAROTROPIC_EOS flag. 

From my own limited understanding of SIMD instructions, I would have assumed embedded if statements would prevent full SIMD vectorisation of these loops. Is the reason that this isn't an issue here because the NON_BAROTROPIC_EOS flag is implemented as a preprocessor instruction and therefore the proper if branch is identifiable at compile time?

Similarly, there are also instances where functions such as std::sqrt and SQR are called within simd blocks. Are these functions already vectorised (declared as simd)? Otherwise I would assume that the compiler would struggle to properly vectorise these blocks. 

Instruction is appreciated, this is a very new area of computer science for me.

**Comment by felker** [comment, marked as answer]
These days, compilers are generally smart enough to auto-vectorize even with runtime conditionals present (tons of caveats to this statement). Certainly if the conditional is a simple constant that has no way to switch from true to false or vice versa within the loop body, it is likely to vectorize without problems. Unfortunately, my impression is that the only ways to be sure are to 1) generate a compiler vectorization report 2) profile or instrument your compiled code. Everything is compiler/version/target architecture/option dependent.

Use vectorized intrinsic functions if you need or want absolute control over the vectorization .

`std::sqrt` should compile to AVX vectorized functions `_mm512_sqrt_pd` or `_mm512_sqrt_ps`, see for example https://www.felixcloutier.com/x86/sqrtps or https://www.intel.com/content/www/us/en/docs/cpp-compiler/developer-guide-reference/2021-8/intrinsics-for-root-function-operations-512-bit.html

`SQR(x)` is just a preprocessor macro that expands to `x*x`

**Reply by felker** [reply]
also look at Clang/Intel ICX `-Wpass-failed` and GCC `-Wopenmp-simd` compiler warning options

**Reply by hwhitehead** [reply]
Cheers Kyle. It can be hard to tell online what computing problems persist or have been solved by more modern systems. I may tinker with vector intrinsics to hopefully accelerate some key areas (for me, an expensive equation of state conversion), but generally I suspect letting the compiler do its own thing will be good enough. Thanks again!
